### PR TITLE
Import Pathway

### PR DESCRIPTION
**Pull request by radamson**
Adds the ability to import a pathway from the UI.  Currently there are no checks for ensuring that the uploaded pathway is actually valid.

Screen Shot 2020-10-06 at 9 24 47 AM

### CODE REVIEW

File: src/components/FileImportModal/FileImportModal.tsx
Comment (dehall) [inline src/components/FileImportModal/FileImportModal.tsx:48]: Somehow a couple typos got introduced here.... `aria` -> `aira` here and `types` -> `tpyes` a few lines below

File: src/components/PathwaysProvider.tsx
Comment (dehall) [inline src/components/PathwaysProvider.tsx:51]: Duplicated `if` here. Sometimes you just really want to make extra sure it's not null

File: src/components/FileImportModal/FileImportModal.tsx
Comment (radamson) [reply src/components/FileImportModal/FileImportModal.tsx:48]: oops

File: src/components/PathwaysProvider.tsx
Comment (radamson) [reply src/components/PathwaysProvider.tsx:51]: just wanted to... double check

File: src/components/CriteriaList/CriteriaList.tsx
Comment (blangley28) [inline src/components/CriteriaList/CriteriaList.tsx:12]: Should these be absolute imports?

File: src/components/PathwaysList/PathwaysList.tsx
Comment (blangley28) [inline src/components/PathwaysList/PathwaysList.tsx:45]: Use a callback here instead of passing in an anonymous function (https://mitrecorp.slack.com/archives/CLH34KL6A/p1594649905003300)

File: src/components/PathwaysList/PathwaysList.tsx
Comment (blangley28) [inline src/components/PathwaysList/PathwaysList.tsx:63]: Use callback here too

**Review by blangley28 — COMMENTED** [review]
I know you're still working on the criteria stuff but I just wanted to make sure you were aware of this - when I import the pathway the criteria dropdown is not selected. The bottom criteria (Male) was built. In this situation are you planing on importing it as a new criteria?
Screen Shot 2020-10-08 at 12 45 05 PM

File: src/components/CriteriaList/CriteriaList.tsx
Comment (radamson) [reply src/components/CriteriaList/CriteriaList.tsx]: They..._absolutely_ should be.

File: src/utils/cpg.ts
Comment (blangley28) [inline src/utils/cpg.ts:80]: Two comments about this:
1) Everything else in this file is specific to CPG. It seems a little weird to create a new `CPGExporter` instance and then export to something other than CPG.
2) There was already the `exportPathway` method in `utils/builder.tsx` which can be modified

File: src/types/pathways-model.d.ts
Comment (blangley28) [inline src/types/pathways-model.d.ts:11]: Is there a reason this was made to be an array? I can't find anywhere it is being used as an array and I'm worried it may have some negative effects on the CPG exporter which only allows a single library to be referenced here

**Review by blangley28 — COMMENTED** [review]
Didn't realize you were still working on the task so I reviewed again - I think these two comments are still worth noting for now though (just in case you were planning on building off them later)

File: src/types/pathways-model.d.ts
Comment (radamson) [reply src/types/pathways-model.d.ts:11]: Turned this into an array so that all the CQL Libraries can be exported too.  If a pathway uses the "TNM" and "Tumor_Markers" Criteria Libraries then they'll be exported in with the pathway in this array.

Do you have a suggestion for how to include them otherwise?  It seems like the model (at least the exported representation) will need to change somehow.

File: src/utils/cpg.ts
Comment (radamson) [reply src/utils/cpg.ts:80]: I was just working in this class for now because I was using some of its features (e.g. it gathers up all the criteria libraries).  Ultimately I think this class would have become more general or it would have moved.  I'll take a look at the version in `utils/builder.tsx` though...didn't realize that was there 👀 .

Comment (radamson) [reply src/utils/cpg.ts:80]: I'll keep this in mind though and rework it.

**Comment by dehall** [comment]
Discussed with @radamson this morning, and we decided on 4 tasks we still need to address:

 - [ ] import any referenced libraries into the CriteriaProvider
 - [ ] re-assign transition CQL/criteriaSource after importing the referenced libraries
 - [ ] the "main" library should only include CQL that was constructed with the criteria builder (anything else should be referenced directly in its original library from the `transition.condition..cql` field )
 - [ ] clean up the cpg.ts logic and make it available to both the CPG exporter and Camino exporter

File: src/utils/cpg.ts
Comment (dehall) [reply src/utils/cpg.ts]: I updated this so there's a new exporter class for the Camino model. Eventually I think we should extract the logic into 3 classes - CPGExporter, CaminoExporter, and common, but I'd rather wait until after Keeyan's PR is merged to avoid a potentially painful rebase.

File: src/types/pathways-model.d.ts
Comment (dehall) [reply src/types/pathways-model.d.ts:11]: Rather than either type, I'd suggest we just make it an array. IMO it's easier to deal with an array of 1 entry rather than check if it's an array or string everywhere.

Comment (radamson) [reply src/types/pathways-model.d.ts]: Made library just an array in 6d7ea8b.

File: src/utils/builder.ts
Comment (blangley28) [inline src/utils/builder.ts:71]: Do we want the CaminoExporter to consider pathway reference nodes? I don't think that should be part of this task but perhaps a back log item to download all of them in a zip file

**Review by blangley28 — COMMENTED** [review]
Looking really good. Was able to create a pathway, export it, and then re-import it again. Two general additional comments:
1. The cql libraries are added as new criteria. If its not too difficult perhaps a checksum can be done on the contents and if its the same as an existing criteria it links to that instead.
2. For the 3 pathways in public/static/pathways the library property needs to be updated to be a list instead of a string

File: src/utils/CaminoExporter.ts
Comment (dehall) [reply src/utils/CaminoExporter.ts:18]: I think this was just a copy & paste from the CPGExporter that I forgot to clean up, will delete unless @radamson gets there first

File: src/utils/builder.ts
Comment (dehall) [reply src/utils/builder.ts:71]: Yeah let's make that a separate task. I was talking to Reece about that on slack:
> The pathway -> cpg translation creates new IDs at export time, so if you exported A then B to CPG separately, any reference from A to B would be lost, so we have to export them together. Whereas with the internal data model it only links by name, so if you export A then B, the relationship is preserved.

Comment (blangley28) [reply src/utils/builder.ts:71]: Created [PATHWAYS-381](https://jira.mitre.org/browse/PATHWAYS-381)

File: src/utils/CaminoExporter.ts
Comment (radamson) [reply src/utils/CaminoExporter.ts]: deleted the TODO

File: src/utils/CaminoExporter.ts
Comment (dehall) [inline src/utils/CaminoExporter.ts:50]: Technically I think there's a chance of false positives here, but more importantly I would have expected the importer to remove the library id for consistency. Nothing else breaks when it's present after importing?

File: public/static/pathways/her2_pathway.json
Comment (blangley28) [inline public/static/pathways/her2_pathway.json:5]: Should this just be an empty list instead?
```suggestion
   "library": [],
```
Same goes for neoadjuvant_pathway.json

File: src/utils/CaminoExporter.ts
Comment (radamson) [reply src/utils/CaminoExporter.ts:50]: I didn't see anything else break because most of the other functionality is keyed off `criteriaSource` as far as I know.  I'll make the change you suggested.

File: public/static/pathways/her2_pathway.json
Comment (radamson) [reply public/static/pathways/her2_pathway.json]: Good call, changed.

File: src/utils/CaminoExporter.ts
Comment (dehall) [reply src/utils/CaminoExporter.ts:50]: After talking with Reece, it seems strange that nothing breaks, and apparently it's a lot easier to do this than to remove it as part of the import process, so in the interest of getting this in soon, we'll leave this as is

Comment (radamson) [reply src/utils/CaminoExporter.ts:50]: So I started working on changing this, but its a little trickier than I expected.

Right now re-associating the CQL requires that the library be prefixed onto the CQL so that it can find the criteria that matches the library/statement combo.  All the existing sample pathways are organized this way too.  I think we'd need to modify the data model to store the information linking the CQL to the library otherwise.  This information has to left on the pathway and I can't have the importer remove it because the re-association is done in a `useEffect` hook after the fact because setting the new criteria is asynchronous.

**Review by mdube7 — APPROVED** [review]
Looks good to me

**Review by blangley28 — COMMENTED** [review]
One last change - if the same pathway is uploaded again (ie same id) it should not upload the pathway and alert the user (I'm fine with a JS alert) that the pathway has not been imported because it already exists.

**Review by blangley28 — APPROVED** [review]

**Pull request merged**
